Add OpenTitan watchdog support

tkng-rivos opened the pull request:
The OpenTitan contains a HWIP called the Always-On (AON) Timer, which has two features - a wakeup timer and a watchdog timer. This patch series is concerned solely with the watchdog feature and does the following:
* Adds the OpenTitan watchdog dts binding
* Adds a driver for the watchdog
* Adds the watchdog for the OpenTitan Earlgrey "board".

More details in the commits.

The AON Timer spec can be found here:
https://opentitan.org/book/hw/ip/aon_timer/index.html

To make the watchdog interrupt work with OpenTitan Earlgrey, some changes were needed to be made to the PLIC configuration.

tkng-rivos committed:
dts: bindings: watchdog: Add OpenTitan AON Timer binding

The OpenTitan AON Timer is a hardware device that has two features:
the wakeup timer and watchdog timer. This commit series implements the
watchdog feature.

The spec can be found here:
https://opentitan.org/book/hw/ip/aon_timer/index.html

Signed-off-by: Tyler Ng <tkng@rivosinc.com>

tkng-rivos committed:
drivers: watchdog: Add the OpenTitan watchdog driver

The OpenTitan watchdog driver is a watchdog driver that can be configured
with two stages: a watchdog "bark", which generates an interrupt, and a
watchdog "bite", which resets the system. The two-stage watchdog can be
enabled by setting CONFIG_WDT_MULTISTAGE=y. Otherwise, the driver
functions as a single-stage watchdog.

A callback function may be set for the bark interrupt through the
wdt setup interface, but will only be used if the two-stage watchdog is
enabled. It must be configured for the first watchdog stage.

The driver controls only the watchdog portion of the OpenTitan AON timer.

Signed-off-by: Tyler Ng <tkng@rivosinc.com>

tkng-rivos committed:
boards: riscv: opentitan_earlgrey: Add the OpenTitan AON Timer

Adds the AON Timer device in the OpenTitan Earlgrey device tree.
Adds overlay files to enable the watchdog and set the alias to
`watchdog0`.

Adds the AON timer (watchdog part) to the supported features section
of the OpenTitan documentation.

Signed-off-by: Tyler Ng <tkng@rivosinc.com>

tkng-rivos committed:
soc: riscv: riscv-privilege: opentitan: Fix PLIC interrupts

1. Fixes the number of interrupts in OpenTitan by default. This should
   be 32 + 185 = 217 IRQs, as there are 185 configurable registers,
   including interrupt 0.

2. Adds 2ND_LVL_INTR_00_OFFSET Kconfig, which is needed to generate a
   PLIC interrupt on IRQ 11.

Signed-off-by: Tyler Ng <tkng@rivosinc.com>

carlocaione commented on drivers/watchdog/wdt_opentitan.c:
```diff
@@ -0,0 +1,178 @@
+	const struct wdt_ot_aontimer_cfg *const cfg = dev->config;
+	volatile struct ot_aontimer_regs *regs = REG_OT_WDT(dev);
+
+	regs->wdog_count = 0;
+	regs->wdog_ctrl |= BIT(0);
```
In general it is preferred to use the `sys_write` / `sys_read` helpers to access and work with MMIO registers. This is done because usually the helpers are also using memory barriers to ensure correctness.

carlocaione commented on drivers/watchdog/wdt_opentitan.c:
```diff
@@ -0,0 +1,178 @@
+static struct wdt_ot_aontimer_data ot_aontimer_data;
+
+static struct wdt_ot_aontimer_cfg ot_aontimer_cfg = {
+	.regs = (struct ot_aontimer_regs *)DT_INST_REG_ADDR(0),
```
you will have always one and only one instance of this?

tkng-rivos replied:
I believe so, yes. I don't think a board will have more than one AON Timer instantiation.

carlocaione replied:
right, but why do not simply use `DT_INST_FOREACH_STATUS_OKAY` as usual? That would be more generic than hardcoding one singleton.

tkng-rivos replied:
I modeled the driver after other watchdog drivers, which seem to hardcode the single instance. Generalizing it would make the implementation more complicated because of the interrupt feature.

carlocaione replied:
@gmarull @mbolivar do we have any strong indication about this?

gmarull replied:
It's ok to have singletons, as soon as driver code doesn't use DT macros directly inside.

gmarull commented on drivers/watchdog/Kconfig.opentitan:
```diff
@@ -0,0 +1,15 @@
+	depends on SOC_RISCV_OPENTITAN
+	depends on DT_HAS_LOWRISC_OPENTITAN_AONTIMER_ENABLED
+	default y
+	select HAS_WDT_MULTISTAGE
+	select WDT_MULTISTAGE
```
why select both? This looks redundant to me

tkng-rivos replied:
`WDT_MULTISTAGE` depends on `HAS_WDT_MULTISTAGE`. It doesn't allow me to select `WDT_MULTISTAGE` directly without selecting the former, and the multistage feature uses `CONFIG_WDT_MULTISTAGE`. I'm not sure if there's a more appropriate way to handle this

tkng-rivos replied:
It seems `HAS_WDT_MULTISTAGE` isn't used anywhere, and I'm not sure what the original intent/purpose of it is. If it is safe to remove I will add a patch to do so, to be included in this series.

tkng-rivos replied:
In this case, I am writing a driver that depends on the `WDT_MULTISTAGE` feature, so it is necessary to select both.

tkng-rivos replied:
Can I resolve this? After a bit of thinking I do have an idea of turning the multi-stage component optional, if you'd like: it would involve skipping the interrupt step and going straight into the reset.

carlocaione replied:
if you think you can resolve this go ahead. Just mark this PR DNM until this is ready again.

gmarull replied:
Isn't this a case where driver _can_ support WDT_MULTISTAGE? Because that means driver should just `select HAS_WDT_MULTISTAGE` and support the feature by adding code conditional to `WDT_MULTISTAGE`, and then, clients will select `WDT_MULTISTAGE` if they need it. Am I missing something? @martinjaeger

tkng-rivos replied:
Alright, I changed the feature.

gmarull commented on drivers/watchdog/Kconfig.opentitan:
```diff
@@ -0,0 +1,15 @@
+
+config WDT_OPENTITAN
+	bool "OpenTitan Always-On (AON) Timer"
+	depends on SOC_RISCV_OPENTITAN
```
redundant

tkng-rivos replied:
Fixed

gmarull commented on drivers/watchdog/wdt_opentitan.c:
```diff
@@ -0,0 +1,179 @@
+
+static int ot_aontimer_init(const struct device *dev)
+{
+	IRQ_CONNECT(DT_IRQ_BY_IDX(NODE_AON, 0, irq),
+		DT_IRQ_BY_IDX(NODE_AON, 0, priority), wdt_ot_isr,
+		DEVICE_DT_INST_GET(0), 0);
+	irq_enable(DT_IRQ_BY_IDX(NODE_AON, 0, irq));
```
please use instance based macros

tkng-rivos replied:
Fixed

gmarull commented on drivers/watchdog/wdt_opentitan.c:
```diff
@@ -0,0 +1,179 @@
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <soc.h>
```
soc.h? why?

tkng-rivos replied:
Fixed

gmarull commented on drivers/watchdog/wdt_opentitan.c:
```diff
@@ -0,0 +1,179 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(wdt_opentitan, CONFIG_WDT_LOG_LEVEL);
+
+#define DT_DRV_COMPAT lowrisc_opentitan_aontimer
```
please move to the top, per convention

tkng-rivos replied:
Fixed

gmarull commented on drivers/watchdog/wdt_opentitan.c:
```diff
@@ -0,0 +1,179 @@
+
+#define DT_DRV_COMPAT lowrisc_opentitan_aontimer
+
+#define REG_OT_WDT(dev) \
+	(((const struct wdt_ot_aontimer_cfg *const)(dev)->config)->regs)
+#define NODE_AON DT_NODELABEL(aontimer)
```
please avoid mixing nodelabel/inst in the same code

gmarull commented on drivers/watchdog/wdt_opentitan.c:
```diff
@@ -0,0 +1,179 @@
+
+#define DT_DRV_COMPAT lowrisc_opentitan_aontimer
+
+#define REG_OT_WDT(dev) \
+	(((const struct wdt_ot_aontimer_cfg *const)(dev)->config)->regs)
```
please avoid helpers like this one

carlocaione replied:
I don't see the fix.

gmarull commented on drivers/watchdog/wdt_opentitan.c:
```diff
@@ -0,0 +1,179 @@
+	(((const struct wdt_ot_aontimer_cfg *const)(dev)->config)->regs)
+#define NODE_AON DT_NODELABEL(aontimer)
+
+struct ot_aontimer_regs {
+	uint32_t alert_test;
+	uint32_t wkup_ctrl;
+	uint32_t wkup_thold;
+	uint32_t wkup_count;
+	uint32_t wdog_regwen;
+	uint32_t wdog_ctrl;
+	uint32_t wdog_bark_thold;
+	uint32_t wdog_bite_thold;
+	uint32_t wdog_count;
+	uint32_t intr_state;
+} __packed;
```
why not use offset definitions?

tkng-rivos replied:
I believe this definition is more clear than using offsets.

carlocaione replied:
Is it though? Like, can you tell straight away what is the offset of the `wdog_count` register?

tkng-rivos commented:
I've made a few adjustments to the driver:
1. I added offset labels to the reg block struct. I would still prefer to have the reg block struct instead of going with an alternative.
2. I changed the logic surrounding how a timeout is chosen when the requested timeout values approach max_uint32. Before, I had simply returned an errored state when the requested max timeout was too large to fit in the hardware register. I added a check to see if it can fit in the requested [min, max] while respecting the bite timeout >= bark timeout requirement.
3. I originally had a `lock` attribute that would be set in the devicetree to indicate that the watchdog configuration would be "locked" by hardware, but removed it before the PR as I thought it wouldn't be so useful. Revisiting it, I added it back because I couldn't see a good way to give the user an option to adjust this setting.

carlocaione commented on drivers/watchdog/wdt_opentitan.c:
```diff
@@ -0,0 +1,201 @@
+	const struct wdt_ot_aontimer_cfg *const cfg = dev->config;
+	volatile struct ot_aontimer_regs *regs = cfg->regs;
+
+	sys_write32(0, (mem_addr_t) &regs->wdog_count);
```
Indeed I think that using the offsets you can get rid of the `(mem_addr_t`) cast that is a bit obnoxious.

`sys_write32(0, reg_base + OFFSET)`

tkng-rivos replied:
Made the change.

tkng-rivos committed:
dts: bindings: power: Add binding for the OpenTitan power manager

The OpenTitan power manager is responsible for changing the OpenTitan's
operation to and from low power state. This patch adds a simple binding
for the power manager's config registers.

This is part of the OpenTitan watchdog patch series. The power manager
HWIP block needs to be configured to enable the watchdog reset
functionality in the OpenTitan Verilator simulation.

Signed-off-by: Tyler Ng <tkng@rivosinc.com>

tkng-rivos committed:
dts: riscv: lowrisc: Add pwrmgr node to OpenTitan Earlgrey devicetree

Adds the pwrmgr devicetree node. This is a simple binding that holds
only the address of the registers for now.

This patch is part of the OpenTitan watchdog (AON Timer) support patch
series. It is needed to ensure that the watchdog reset functionality
is enabled.

Signed-off-by: Tyler Ng <tkng@rivosinc.com>

tkng-rivos committed:
soc: riscv: riscv-privilege: opentitan: Enable watchdog reset function

Adds watchdog reset as a reset source at SoC init. This is achieved by:

1. Writing 0x2 to the RESET_EN bitfield register to indicate watchdog
reset is enabled.
2. Writing 0x1 to the CFG_CDC_SYNC register to commit the change.
3. Polling the CFG_CDC_SYNC register until reading 0 to confirm the
change has been processed.

This patch is part of the OpenTitan watchdog (AON Timer) support patch
series. It is needed to ensure that the watchdog reset functionality
is enabled. Note that the timer itself is not enabled here, only the
reset function.

Signed-off-by: Tyler Ng <tkng@rivosinc.com>

tkng-rivos commented:
I made a mistake when opening the PR in that I did not test it with the OpenTitan Verilator simulation; instead I used a QEMU setup I have locally. Lo and behold, it did not work with the Verilator simulation. I have investigated the cause and it seems like some additional configuration was needed to activate the watchdog reset functionality in the power management block.

Thus, I have included additional patches for adding a barebones devicetree binding for the power manager register block in the OpenTitan Earlgrey. All it has is a register configuration so that we can define the addresses of the necessary registers are; nothing else is implemented. These registers are used at SoC init to enable the watchdog reset. Note that the watchdog is not active until the watchdog timer itself is activated. Without these patches, the watchdog "bite" will never trigger a reset.

I have since tested the Verilator simulation and confirmed that both the bark and bite work with a modified version of the watchdog sample, which basically checks for the `WDT_MULTISTAGE` configuration option and adjusts the watchdog timeout installation with a 2-stage timeout. If interested, I can add that change to the patch series.

fkokosinski reviewed (dismissed):
LGTM

carlocaione commented on drivers/watchdog/wdt_opentitan.c:
```diff
@@ -0,0 +1,218 @@
+#define OT_REG_INTR_STATE_OFFSET 0x24
+
+struct wdt_ot_aontimer_cfg {
+	volatile uintptr_t regs;
```
no need for this to be volatile

gmarull replied:
+1

gmarull commented on boards/riscv/opentitan_earlgrey/opentitan_earlgrey_defconfig:
```diff
@@ -9,3 +9,4 @@ CONFIG_SERIAL=y
 CONFIG_CONSOLE=y
 CONFIG_UART_CONSOLE=y
 CONFIG_BUILD_OUTPUT_BIN=n
+CONFIG_WDT_MULTISTAGE=n
```
this looks redundant

gmarull commented on soc/riscv/riscv-privilege/opentitan/soc.c:
```diff
@@ -8,10 +8,18 @@
 
 static int soc_opentitan_init(void)
 {
+	/* Enable the watchdog reset (bit 1). */
+	sys_write32(2u, PWRMGR_BASE + PWRMGR_RESET_EN_REG_OFFSET);
+	/* Write CFG_CDC_SYNC to commit change. */
+	sys_write32(1u, PWRMGR_BASE + PWRMGR_CFG_CDC_SYNC_REG_OFFSET);
+	/* Poll CFG_CDC_SYNC register until it reads 0. */
+	while (sys_read32(PWRMGR_BASE + PWRMGR_CFG_CDC_SYNC_REG_OFFSET)) {
+	}
```
question: why isn't this handled by the driver itself?

tkng-rivos replied:
Adding the watchdog to the SoC's list of enabled reset sources seems more appropriate to be in the SoC init code.

tkng-rivos commented:
Fixed some dead links in the OpenTitan Earlgrey page as well.

fkokosinski commented:
@tkng-rivos could you rebase?

carlescufi commented:
@tkng-rivos please rebase and resolve conflicts
